### DISCUSSION TITLE
What is the best way to access all users from the auth.users table?

### DISCUSSION THREAD

**Discussion by RafaelCENG** [supabase/supabase#18780] [Questions] [answered]
Hello everyone I'm trying to access my auth.users table from the client side. I have an admin account where I want to use data like email, last login, etc. I do not want to create a new table based on auth.users. Is my best option to use a view? Also, I tried something like this:
```
    console.log(supabase.auth.admin.listUsers().then((test) => test.data));
```
Which is wrong based on the docs. (Also getting unauthorized error)
I managed to get my current user with getSession() but I need the whole table.

**Comment by GaryAustin1** [comment, marked as answer]
The API has no access to the auth schema.   You might be able to use a view.  A security definer function called with rpc would work.  The issue with both approaches is you will need to protect the view or rpc call so a normal user can’t call it.

**Reply by ChronSyn** [reply]
Is this a recent change? It used to be possible to use a Supabase client with `db.schema` set to `auth`, and you could then perform queries. Useful from a backend perspective, especially for scenarios where you want to look up a user (e.g. by email, etc).

(I ask this because I've just started having trouble with it, and if that's due to a change, then that's a significant breaking change which I haven't seen mentioned anywhere - in docs, github, or relevant social media channels)

**Reply by GaryAustin1** [reply]
It has Never been possible to access the auth schema from the REST API.  It is not allowed in the API schema settings and none of the API roles have grants on it.
All you can do is use security definer functions to access it from an rpc call.

Edit.  The database port should have access to it if that is what you mean, but that is not considered an API.

**Reply by TangSirOnGit** [reply]
As an application administrator, if I want to list the number of registered users on the admin dashboard，

I attempted to retrieve user data using the following method from the documentation:

```
const {
    data: { users },
    error,
  } = await supabase.auth.admin.listUsers({
    page: 1,
    perPage: 1000,
  });
```
However, the resulting user list is empty.

**Reply by GaryAustin1** [reply]
@TangSirOnGit You should check the error returned.  Is your client initialized with service_role?

**Reply by TangSirOnGit** [reply]
Hey, @GaryAustin1 thank you for reply,
the error message:

users: []
error: AuthApiError: User not allowed
    at handleError (webpack-internal:///(rsc)/./node_modules/@supabase/gotrue-js/dist/module/lib/fetch.js:46:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async _handleRequest (webpack-internal:///(rsc)/./node_modules/@supabase/gotrue-js/dist/module/lib/fetch.js:93:9)
    at async _request (webpack-internal:///(rsc)/./node_modules/@supabase/gotrue-js/dist/module/lib/fetch.js:73:18)
    at async GoTrueAdminApi.listUsers (webpack-internal:///(rsc)/./node_modules/@supabase/gotrue-js/dist/module/GoTrueAdminApi.js:153:30)
    at async Component (webpack-internal:///(rsc)/./app/admin/(overview)/page.jsx:35:40) {
  __isAuthError: true,
  status: 401
}

how to initialize the client with 'service_role' ?

thank you!

**Reply by Donald646** [reply]
No I'm pretty sure this is cause you need the admin client.

**Reply by rindev0901** [reply]
@TangSirOnGit You can implement service_role like this below code
```
export function createClient({ admin }: Props = { admin: false }) {
  const cookieStore = cookies();

  if (admin) {
    const env = process.env.NODE_ENV;
    return createServerClient<Database>(
      process.env.NEXT_PUBLIC_SUPABASE_URL!,
      env === "development"
        ? process.env.NEXT_PRIVATE_SUPABASE_SERVICE_ROLE_KEY_DEVELOPMENT!
        : process.env.NEXT_PRIVATE_SUPABASE_SERVICE_ROLE_KEY_PRODUCTION!,
      {
        cookies: {
          getAll() {
            return [];
          },
        },
      }
    );
  }

  return createServerClient<Database>(
    process.env.NEXT_PUBLIC_SUPABASE_URL!,
    process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
    {
      cookies: {
        getAll() {
          return cookieStore.getAll();
        },
        setAll(cookiesToSet) {
          try {
            cookiesToSet.forEach(({ name, value, options }) =>
              cookieStore.set(name, value, options)
            );
          } catch {
            // The `setAll` method was called from a Server Component.
            // This can be ignored if you have middleware refreshing
            // user sessions.
          }
        },
      },
    }
  );
}

```

**Comment by dilipBadal** [comment]
Hi, I have the same problem but its not consistent. Once every 2 tries, it works but that 1 time it says user not allowed.

This is how I'm doing,

```
supabase = create_async_client(
            supabase_url=supabase_url,
            supabase_key=service_role_key
)


Which I use it later as this

data = await supabase.auth.admin.listUsers()
```




And when I check my auth logs, it says  'this token needs to have one of the following roles: supabase_admin, service_role'

The success rate of this is 50/50. If it was failing always I could find it but for some reason I cannot seem to figure it our.
